chalk-ir/src/zip.rs

```diff
@@ -324,21 +452,184 @@ impl<I: Interner> Zip<I> for VariableKind<I> {
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        Zip::zip_with(zipper, variance, &a.opaque_ty_id, &b.opaque_ty_id)?;
+        zipper.zip_substs(
+            None,
+            a.substitution.parameters(interner),
+            b.substitution.parameters(interner),
+        )
+    }
+}
+
+impl<I: Interner> Zip<I> for ApplicationTy<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
     where
         I: 'i,
     {
         let interner = zipper.interner();
-        Zip::zip_with(zipper, a.data(interner), b.data(interner))
+        Zip::zip_with(zipper, variance, &a.name, &b.name)?;
+        match a.name {
+            TypeName::FnDef(fn_def_id) => zipper.zip_substs(
+                Some(zipper.unification_database().fn_def_variance(fn_def_id)),
+                a.substitution.parameters(interner),
+                b.substitution.parameters(interner),
+            ),
+            TypeName::Adt(adt_id) => zipper.zip_substs(
+                Some(zipper.unification_database().adt_variance(adt_id)),
+                a.substitution.parameters(interner),
+                b.substitution.parameters(interner),
+            ),
+            TypeName::Ref(mutbl) => {
+                Zip::zip_with(
+                    zipper,
+                    Variance::Contravariant,
+                    &a.substitution.parameters(interner)[0],
+                    &b.substitution.parameters(interner)[0],
+                )?;
+                let variance = match mutbl {
+                    Mutability::Not => Variance::Covariant,
+                    Mutability::Mut => Variance::Invariant,
+                };
+                Zip::zip_with(
+                    zipper,
+                    variance,
+                    a.substitution.iter(interner).last().unwrap(),
+                    b.substitution.iter(interner).last().unwrap(),
+                )?;
+                Ok(())
+            }
+            _ => zipper.zip_substs(
+                None,
+                a.substitution.parameters(interner),
+                b.substitution.parameters(interner),
+            ),
+        }
+    }
+}
+
+impl<I: Interner> Zip<I> for DynTy<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        Zip::zip_with(zipper, variance, &a.bounds, &b.bounds)?;
+        Zip::zip_with(zipper, Variance::Contravariant, &a.lifetime, &b.lifetime)?;
+        Ok(())
+    }
+}
+
+impl<I: Interner> Zip<I> for FnSubst<I> {
+    fn zip_with<'i, Z: Zipper<'i, I>>(
+        zipper: &mut Z,
+        _variance: Variance,
+        a: &Self,
+        b: &Self,
+    ) -> Fallible<()>
+    where
+        I: 'i,
+    {
+        let interner = zipper.interner();
+        zipper.zip_substs(
+            None,
```

Comment:
These are the fn parameters, right? If so, we should comment, but also this should be contravariant.

Reply:
Well, actually, it should be `variance.xform(Contravariant)`